helix-term/src/ui/context.rs

```diff
@@ -0,0 +1,510 @@
+#[allow(deprecated)]
+use helix_core::visual_coords_at_pos;
+
+use helix_core::{
+    syntax::RopeProvider,
+    text_annotations::TextAnnotations,
+    tree_sitter::{Node, QueryCursor, QueryMatch},
+    Position,
+};
+
+use helix_view::{editor::Config, graphics::Rect, Document, Theme, View, ViewId};
+
+use tui::buffer::Buffer as Surface;
+
+use crate::ui::text_decorations::DecorationManager;
+
+use super::{
+    document::{render_text, TextRenderer},
+    EditorView,
+};
+
+#[derive(Debug, Default, Clone)]
+pub struct StickyNode {
+    pub line: usize,
+    pub visual_line: u16,
+    pub byte_range: std::ops::Range<usize>,
+    pub indicator: Option<String>,
+    pub anchor: usize,
+    pub has_context_end: bool,
+    pub view_id: ViewId,
+}
+
+#[derive(Debug)]
+struct StickyNodeContext {
+    /// The row on which the cursor is placed
+    pub visual_row: usize,
+    /// This marks the location of which we take possible out of range nodes
+    /// e.g. on follows-cursor: 'on', this would be the parent nodes, relative to the *cursor*
+    /// on follows-cursor: 'off', this would be the parent nodes, relative to the *topmost screen*
+    pub context_location: usize,
+    /// The topmost byte that is visible (and not hidden behind sticky nodes)
+    pub topmost_byte: usize,
+    /// The anchor of the view offset
+    pub anchor_line: usize,
+    /// The viewport
+    pub viewport: Rect,
+}
+
+impl StickyNodeContext {
+    pub fn from_context(
+        last_nodes: &Option<Vec<StickyNode>>,
+        doc: &Document,
+        view: &View,
+        config: &Config,
+        cursor_cache: &Option<Position>,
+    ) -> Option<Self> {
+        let Some(cursor_cache) = cursor_cache else {
+            return None;
+        };
+        let text = doc.text().slice(..);
+        let viewport = view.inner_area(doc);
+        let cursor_byte = text.char_to_byte(doc.selection(view.id).primary().cursor(text));
+
+        let anchor_line = text.char_to_line(view.offset.anchor);
+        let visual_cursor_row = cursor_cache.row;
+
+        if visual_cursor_row == 0 {
+            return None;
+        }
+
+        let top_first_byte = text
+            .try_line_to_byte(anchor_line + last_nodes.as_ref().map_or(0, |nodes| nodes.len()))
+            .ok()?;
+
+        let last_scan_byte = if config.sticky_context.follow_cursor {
+            cursor_byte
+        } else {
+            top_first_byte
+        };
+        Some(Self {
+            visual_row: visual_cursor_row,
+            context_location: last_scan_byte,
+            topmost_byte: top_first_byte,
+            anchor_line,
+            viewport,
+        })
+    }
+}
+
+/// Calculates the sticky nodes
+pub fn calculate_sticky_nodes(
+    nodes: &Option<Vec<StickyNode>>,
```

Comment:
```suggestion
    nodes: Option<&Vec<StickyNode>>,
```
Inspired by https://github.com/helix-editor/helix/pull/11091